api/certs/cert.go

```diff
@@ -0,0 +1,218 @@
+)
+
+type certCache struct {
+	Domain            string `json:"domain"`
+	CertURL           string `json:"certUrl"`
+	CertStableURL     string `json:"certStableUrl"`
+	PrivateKey        []byte `json:"privateKey"`
+	Certificate       []byte `json:"certificate"`
+	IssuerCertificate []byte `json:"issuerCertificate"`
+	CSR               []byte `json:"csr"`
+}
+
+func (c *certCache) toResource() *certificate.Resource {
+	return &certificate.Resource{
+		Domain:            c.Domain,
+		CertURL:           c.CertURL,
+		CertStableURL:     c.CertStableURL,
+		PrivateKey:        c.PrivateKey,
+		Certificate:       c.Certificate,
+		IssuerCertificate: c.IssuerCertificate,
+		CSR:               c.CSR,
+	}
+}
+
+func VerifyOCSP(bundle []byte) (bool, error) {
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return false, err
+	}
+
+	myUser := User{
+		Key: privateKey,
+	}
+
+	c := lego.NewConfig(&myUser)
+
+	// A client facilitates communication with the CA server.
+	client, err := lego.NewClient(c)
+	if err != nil {
+		log.Error(err)
+		return false, err
+	}
+	_, oscp, err := client.Certificate.GetOCSP(bundle)
+	if err != nil {
+		return false, err
+	}
+	if oscp.Status == 0 {
+		return true, nil
+	}
+	return false, fmt.Errorf("cert did not pass OCSP check")
+}
+
+func VerifyCertificate(certPath string, domain string) (bool, error) {
+	bundle, err := ioutil.ReadFile(certPath)
+	if err != nil {
+		return false, err
+	}
+	unRevoked, err := VerifyOCSP(bundle)
+	if !unRevoked {
+		return false, err
+	}
+	certs, err := certcrypto.ParsePEMBundle(bundle)
+	if err != nil {
+		return false, err
+	}
+	if certs[0].Subject.CommonName != domain {
+		return false, fmt.Errorf("inconsistent domain names")
+	}
+	return true, nil
+}
+
+func VerifyLocalCertificate(certPath string, keyPath string, domain string) (bool, error) {
+	if len(certPath) > 0 && len(keyPath) > 0 && len(domain) > 0 && common.FileExisted(certPath) && common.FileExisted(keyPath) {
+		return VerifyCertificate(certPath, domain)
+	}
+	return false, nil
+}
+
+func CertExists() (bool, bool, bool) {
+	wssUserCertExists, err := VerifyLocalCertificate(config.Parameters.HttpWssCert, config.Parameters.HttpWssKey, config.Parameters.HttpWssDomain)
+	if err != nil {
+		log.Error(err)
+	}
+	httpsUserCertExists, err := VerifyLocalCertificate(config.Parameters.HttpsJsonCert, config.Parameters.HttpsJsonKey, config.Parameters.HttpsJsonDomain)
+	if err != nil {
+		log.Error(err)
+	}
+	defaultDomain, err := node.GetDefaultDomainFromIP(config.Parameters.Hostname, config.Parameters.DefaultTlsDomainTmpl)
+	if err != nil {
+		log.Error(err)
+	}
+	_, _, certFile, keyFile := GetACMEFileNames(defaultDomain, config.Parameters.CertDirectory)
+	defaultCertExists, err := VerifyLocalCertificate(certFile, keyFile, defaultDomain)
+	if err != nil {
+		log.Error(err)
+	}
+
+	return wssUserCertExists, httpsUserCertExists, defaultCertExists
+}
+
+func CheckCerts() (chan struct{}, chan struct{}) {
```

Comment:
Suggest to rename to PrepareCerts, and rename the current PrepareCerts to prepareCerts